File: past/20200124_47/Alex.scala
object alex {
  def permuteUnique(nums: Array[Int]): List[List[Int]] = {
    import scala.collection.mutable
    var res: mutable.Stack[List[Int]] = mutable.Stack()
    var t: mutable.Stack[Int] = mutable.Stack()
    val used = Array.ofDim[Boolean](nums.length)

    def dfs(candidates: Array[Int], s: Int, temp: mutable.Stack[Int], res: mutable.Stack[List[Int]], used: Array[Boolean]): Unit = {
      if (temp.length == candidates.length) {
        res push temp.toList
        return
      }
      var i = 0
      while (i < candidates.length) {
        if (used(i) || i > 0 && candidates(i) == candidates(i - 1) && !used(i - 1)) {
        } else {
          used(i) = true
          temp.push(candidates(i))
          dfs(candidates, i, temp, res, used)
          used(i) = false
          temp.pop()
        }
        i += 1
      }
    }

    val tt = nums.sorted
    dfs(tt, 0, t, res, used)
    res.toList
  }
}

File: past/202001/20200122_31/Alex.java
class Alex{
    public  void nextPermutation(int[] nums){
        for(int i = nums.length - 2;i >= 0;i--){
            if(nums[i] < nums[i + 1]){
                for(int j = nums.length - 1;j > i ; j--){
                    if(nums[j] > nums[i]){
                        int tmp = nums[j];
                        nums[j] = nums[i];
                        nums[i] = tmp;
                        break;
                    }
                }
                reverseArray(nums,i + 1,nums.length - 1);
                return;
            }


        }
        reverseArray(nums,0,nums.length - 1);
    }

    private   void reverseArray(int[] nums, int begin, int end){
        for(int i = begin, j = end;i<j;i++,j--){
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
    }
}